third_party/nvfuser/csrc/scheduler/normalization_utils.cpp

```diff
@@ -494,5 +502,193 @@ std::optional<GridOuterNormalizationParams> getGridOuterNormalizationParams(
   return std::nullopt;
 }
 
+namespace {
+
+// Go through the resolution points one by one. Resolution points are points
+// in which the reduction branch meets the residual branch. These are points
+// where the persitent buffer may no longer be needed (one point could be
+// after another, and the buffer would be needed until the last resolution
+// points)
+std::vector<TensorView*> getPersistentUseOfBuffer(
+    TensorView* buffer,
+    TensorView* resolution_point) {
+  std::vector<TensorView*> persistent_use_of_buffer;
+
+  // Need to go through all paths from the persistent buffer to the
+  // resolution point
+  auto chains_to_resolution =
+      DependencyCheck::getAllDependencyChains(buffer, resolution_point);
+  for (auto chain : chains_to_resolution) {
+    auto tv_chain = ir_utils::filterByType<TensorView>(chain);
+
+    // To move the persistent buffers to the inputs, we need to recompute
+    // the persistent buffer for all branches that don't go through a
+    // reduction. If there's a reduction on the current path between the
+    // persistent buffer and resolution, continue, there's no need to
+    // replicate this use.
+    if (std::any_of(tv_chain.begin(), tv_chain.end(), [](TensorView* tv) {
+          return tv->hasReduction();
+        })) {
+      continue;
+    }
+
+    // Grab use of the buffer, chain[0] is the persistent buffer, chain[1]
+    // is its first use.
+    auto use = chain[1];
+
+    // Only grab unique uses, a persistent buffer could be used multiple
+    // times in the same expression.
+    if (std::find(
+            persistent_use_of_buffer.begin(),
+            persistent_use_of_buffer.end(),
+            use) != persistent_use_of_buffer.end()) {
+      continue;
+    }
+    persistent_use_of_buffer.emplace_back(use->as<TensorView>());
+  }
+
+  return persistent_use_of_buffer;
+}
+
+} // namespace
+
+std::vector<TensorView*> projectPersistentBuffers(Fusion* fusion) {
+  auto persistent_info = scheduler_utils::persistentBuffers(fusion);
+  std::vector<TensorView*> dummy_outputs;
+
+  // Convenience accessors
+  const auto& persistent_buffers = persistent_info.persistent_buffers;
+  const auto& persistent_resolution_points =
+      persistent_info.persistent_buffer_resolution_points;
+  const auto& projected_buffers =
+      persistent_info.projectable_persistent_buffers;
+
+  TORCH_INTERNAL_ASSERT(
+      persistent_buffers.size() == persistent_resolution_points.size());
+
+  // Iterate through projected buffers, tracking which index it corresponds too
+  // since there's a resolution point entry for every buffer.
+  for (auto buffer_i : c10::irange(persistent_buffers.size())) {
+    auto buffer = persistent_buffers[buffer_i];
+    if (std::find(projected_buffers.begin(), projected_buffers.end(), buffer) ==
+        projected_buffers.end()) {
+      continue;
+    }
+
+    for (auto resolution_point : persistent_resolution_points[buffer_i]) {
+      const auto persistent_use_of_buffer =
```

Comment:
`persistent_use_of_buffer` was defined outside of this loop. I believe this is the correct position.

Reply:
Maybe not. Will update.

Reply:
Updated. See #2576